### ISSUE TITLE
docker service create doesn't allow --privileged flag

### ISSUE THREAD

**Issue by frellus** [docker/docker#24862]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 03:39:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 03:39:43 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 54
Server Version: 1.12.0-rc4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 71
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: active
 NodeID: 33ops9juo9ea1twbfq2dyt89y
 IsManager: Yes
 Managers: 2
 Nodes: 5
 CACertHash: sha256:cef0da32ea05dd1038a5b8ae1a3a6956b6a5efa2d2fcad535a696dd568220197
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 3.13.0-86-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 94.42 GiB
Name: irvm-ggallag
ID: WA3H:N54J:H7F3:CQV6:74ZX:IWIZ:U6XG:2VCB:45LP:LDD5:FHB6:7CWZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Ubuntu 14.04 VM under KVM running Docker enginer 1.12 RC4

**Steps to reproduce the issue:**
1. docker service create <name>
2. inside docker image NFS mount

**Describe the results you received:**
I can run "docker run --privileged" to allow an NFS mount from within my container, however there is no way to pass this --privileged flag to "docker service" and if I do not pass the --privileged flag, the contain would error internally on the mount like:
```
mount: permission denied
```

**Describe the results you expected:**
I should be able to have my container mount an NFS server from within it. I do not want to do this externally or via a docker volume, for example, I am trying to drive a huge number of parallel containers running NFS mounts and I/O individually.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by justincormack** [comment]
I think there is a whole set of issues for these features on `service create`, we should probably make an issue listing them all.

I think the plan was to discuss what should be added, once 1.12 is released.

**Comment by thaJeztah** [comment]
Correct, I was planning to create a tracking issue for that

**Comment by gerred** [comment]
I could really use this for 1.12 as well. If this is an area I could jump in and issue a PR, I'm happy to get started on it.

**Comment by thaJeztah** [comment]
We need to decide first; services are not "containers", so not all options can be / should be copied to service create

**Comment by gerred** [comment]
@thaJeztah Another consideration - I have different needs between a replicated service and a global one. If a jobs service type is introduced, which has been discussed, those needs might be different too.

The global one I may expect to have more flags/options around, just given the nature of "other things" I might be doing with them (monitoring, networking, running containers, etc.). I suppose I could have a global service that mounts the docker socket that then run a privileged container on each node, but that seems messy (now my global service lifecycle is managing the lifecycle of a container on each engine separately).

Hopefully that helps with some of that discussion.

**Comment by frellus** [comment]
If services != containers, why do you pass an image name to the create command?  Seems like you would rather pass something a manifest (like the docker-compose.yml file ?).

For this issue, if it is a PR, I'll phrase it in user story format:  As a user of swarm, I want to create services and containers which run under privileged mode. How do I do this?

I'm happy to help any way that I can!

**Comment by gerred** [comment]
@frellus You're right - stacks/DABs are really what I need but they're also really early and don't have a service type option associated with them yet. There's also some other little nits there I need to write a more specific issue around in compose. Ultimately it's all still a bit of a chicken and egg problem - in one I get privileged, in the other I get service types. :) It'll all shake out, for now just reporting my uses to help provide as much data as I can! :smile:

**Comment by padyx** [comment]
I am very interested in this, because as far as I know the Oracle DB cannot run in a container without either the `--privilege` option or specifying the `--shm` flag when running a container.

Given that they are both not supported yet in API 1.24 for services AFAICS, it would be impossible to replace a Docker Swarm (standalone) by Docker 1.12 Swarm to run such services.

**Comment by acaranta** [comment]
Hi ! Indeed I run into an issue with this lack of privileged flag .... we need to be able to deploy CUDA enabled containers .... but without this kind of options, well, mount the device (GPU, etc) will be a big problem .... actually, it will be a nogo ... :(

**Comment by bboreham** [comment]
Drive-by observation: you should do `--cap-add` before `--privileged`, to encourage people to be more granular in what they need.  #25885 relates.

**Comment by b0ch3nski** [comment]
I'd really like to see it implemented soon - there are more solutions that require ``--privileged`` flag to function properly, e.g. **cAdvisor** which I'm using for containers performance monitoring.

**Comment by acaranta** [comment]
If I may add ... `--privileged` and/or `--device*` are quite critical for the case you need to run containers using GPU/CUDA calculations .... 
Placement rules to allow these kind of containers to run on specific hosts can be used ... but not being able to actually use the GPU ..... kind of renders the swarmmode useless for us ... :(

**Comment by AkihiroSuda** [comment]
Just FYI, linking the PR for supporting "device" to this issue:
https://github.com/docker/swarmkit/issues/1244
https://github.com/docker/swarmkit/pull/1355


Even though there would be `--device`,  I think `--privileged` is still attractive (e.g. for DinD)

**Comment by frimdo** [comment]
+1

`--cap-add` and `--cap-drop` Is a must. `--privileged` would be nice. Are there any plans implementing it?

**Comment by vingrad** [comment]
+1

**Comment by fjammes** [comment]
Missing --cap-add to use swarm mode in production, my management push me to move towards kubernetes if this option is not added soon. Do you have some plan and agenda for adding this feature please?

**Comment by dalefwillis** [comment]
+1

`--cap-add` would be a huge help!

**Comment by seiferteric** [comment]
Working on kind of a [workaround](https://github.com/seiferteric/docker-service-proxy). It will run your privileged app in a secondary container by mounting /var/run/docker.sock in your service and proxying tcp connections back to the service container with socat and unix sockets. Still needs some work though.

**Comment by calh** [comment]
I'd also like to see the `--cap-add` on docker services.  I've written a workaround, similar to @seiferteric's if anyone would like to try it out:  https://github.com/calh/docker_priv_proxy

I'm using signal traps, socat, and the docker socket to pair swarm mode service containers with a local privileged mode container.  It seems to work well so far!

**Comment by galindro** [comment]
Any return about this feature guys?

**Comment by getvivekv** [comment]
+1

**Comment by pinkal-vansia** [comment]
+1

**Comment by seiferteric** [comment]
Just another use case, I want to run keepalived with vrrp on a swarm and it needs net=host and --cap-add=NET_ADMIN, so cap-add would be great.

**Comment by joaquin386** [comment]
+1

**Comment by TAGC** [comment]
+1

I want to be able to run a dockerised web service on an RPi that interfaces with a microcontroller over USB and provides web-based access to it over a network. I'm blocked by this and dotnet/dotnet-docker#223.

**Comment by voron3x** [comment]
+1

**Comment by soar** [comment]
Up, need this feature!

**Comment by siavash9000** [comment]
+1

**Comment by Farfad** [comment]
+1

**Comment by 2stacks** [comment]
+1

**Comment by Kent-H** [comment]
+1

**Comment by dfresh613** [comment]
+1

**Comment by spfannemueller** [comment]
+1

**Comment by barhun** [comment]
+1

**Comment by dl00** [comment]
+1

**Comment by wangsquirrel** [comment]
+1

**Comment by AkihiroSuda** [comment]
Folks, please avoid filling up this issue with `+1` 😭 
You could click the 👍 button instead.


<img width="859" alt="a" src="https://cloud.githubusercontent.com/assets/9248427/25328881/7d510326-2914-11e7-866b-90da678a33e3.png">

**Comment by matan-insoundz** [comment]
+1

**Comment by subfuzion** [comment]
Here's a specific (admittedly edge) use case: I distribute DinD containers around a cluster of host nodes (the DinD containers allow us to run mini-isolated test swarms). If I were a able to use privileged services on an outer host swarm, I could take advantage of swarm for the automatic distribution of these workloads (and named access) without having to manage these details manually.

**Comment by realcbb** [comment]
+1

**Comment by Toshik** [comment]
Any news on ```--device``` for services?

**Comment by MingShyanWei** [comment]
+1

**Comment by gklbiti** [comment]
+1

**Comment by cpuguy83** [comment]
@Toshik #33440, it's not complete in terms of actually allocating the device, but putting the right API's in place.

**Comment by danmanners** [comment]
+1

**Comment by blop** [comment]
+1000

**Comment by TAGC** [comment]
![1 year exactly since this issue was opened](http://i.imgur.com/q67UnPu.jpg)

**Comment by Kent-H** [comment]
+9001

**Comment by thaJeztah** [comment]
Please don't `+1` this issue, it's not helpful. If you have specific requirements to need this feature, please _subscribe_ the use case, which is more helpful.

At this point it is not likely that `--privileged` will be added as-is to `docker service`, as this option nullifies all security that containers provide (TL;DR; running a service/container with `--privileged` effectively gives root access to the host).

Instead, more fine-grained permissions are needed; while `--cap-add` / `--cap-drop` is an option, it doesn't solve everything; for example, security profiles (SecComp, SELinux, AppArmor) are also related. Setting the right combination of options is complex (and cumbersome), which is why often people "just set `--privileged`" to make things work (but doing so, running highly insecure - see my earlier comment).

Security should be usable, therefore a proposal is opened to work towards a design that both allows people to give their service the privileges they need in a more fine-grained approach, without introducing the complexity of setting all options manually.

The proposal can be found here: [Proposal: Entitlements in Moby](https://github.com/moby/moby/issues/32801), and the POC https://github.com/docker/libentitlement repository, also there's a [Google document describing the proposal in more detail](https://docs.google.com/document/d/1j3BJUNBsgi-nxJHoIJHsXRRtVWT5lrwsI2EN9WMQaes/edit#heading=h.yhnr195944yh).

We welcome feedback on the proposal; either by commenting on that issue, or taking part in the Orchestration Security SIG meetups (see https://forums.mobyproject.org/c/sig/orchestration-security)


@2stacks
@blop 
@dalefwillis
@danmanners 
@dfresh613 
@dl00
@Farfad
@galindro
@getvivekv
@gklbiti
@joaquin386
@Kent-H
@matansag
@MingShyanWei 
@pinkal-vansia
@realcbb
@siavash9000 
@soar
@spfannemueller 
@TAGC
@vingrad
@voron3x
@wangsquirrel

I removed your `+1` comments, as they don't add useful input to the discussion; if you don't have a particular use-case / additional information that can help getting the design right, use the :+1: emoji in the top comment to let others know you're interested in this (see @AkihiroSuda's comment https://github.com/moby/moby/issues/24862#issuecomment-296577597). If you commented just to get notified on updates; use the **subscribe** button on the right side of this page.

**Comment by elliotwoods** [comment]
Use case : 
Distributed embedded environment with multiple nodes (>100) running headless in the wild. Need to be able to remotely update applications / environments. The embedded systems operate external electrical devices (e.g. smart home / industrial automation). Example devices:

* USB to serial (either the USB or the TTY device could be shared)
* UARTs (TTY)
* GPIO pins
* I2C / SPI
* On-board LED's

Current plan is to run without `docker service` (i.e. 'manual swarm') until a workable solution exists.

**Comment by chestnutt** [comment]
Another use case:
Using a Container-as-a-service environment (so no ability to install anything on the host). We want to deploy an image (--mode=global) that can report back the stats we care about on all other docker containers running on the same host, as well as on the host itself. (Similar to @wannabesrevenge's case )

**Comment by port22** [comment]
I wonder if it's anyhow possible to run a docker container with the privileged option on a swarm node, when you only have access to a swarm master. If not, the argument that a service is not a container is half-viable. of course it isn't, but if we can't run a container on another node, it's not possible to do it anyway.

I mainly need this for bootstrapping/provisioning nodes

**Comment by marcellodesales** [comment]
I'm in desperate needs for this feature to run a `mode: global` in swarm using docker-compose v3. 

My current workaround is to drop the docker-compose.yml and use compose `v2` instead and execute it with `privileged: true`.

**Comment by goalotc** [comment]
To deploy containers such as sysdig. Seccomp, being proposed by Docker but not supported in swarm service is unreasonable.

**Comment by ab77** [comment]
IPSec VPN with strongSwan/Openswan is another use-case for `privileged` support.

**Comment by TAGC** [comment]
I have an ASP.NET Core web service that I've dockerised. Part of the operation of this web service involves communicating with a microcontroller over USB. I can run it on my local Windows 7 machine using Docker Toolbox by passing the USB device through to Virtualbox (so the docker VM has access to it).

On Windows 10 deployment machines it would be preferable to use Docker For Windows, which runs Docker on top of Hyper-V. However, Hyper-V apparently doesn't support USB pass-through like Virtualbox does. One idea I had to work around this would be to split my webservice into two separate ones; one that handles all the USB communication logic and the other handling everything else. That would mean I could run the former (the "device comms" service) on a Raspberry Pi and the latter would be free to run anywhere, even in a non-privileged container. The idea would be to have a Docker swarm with these two services linked to each other.

However, how can I create a "device comms" service which requires **privileged access** to the filesystem if `--privileged` is not supported? How do I make *any* sort of Dockerised application that needs to perform USB communication on a Windows 10 machine?

This issue has been open for 14 months and it doesn't seem like any progress is being made.

**Comment by man4j** [comment]
I want to deploy GlusterFS in swarm, but glusterfs dont work without --privileged flag ((( Please help!

**Comment by purplesrl** [comment]
@thaJeztah 

The proposal looks great, but that looks like a complete change of the way privileges work, which will take a loooooong time. 

Meanwhile docker allows you to run --privileged, or to use --cap-add for containers, so why not provide the same facility for services and let the users decide if they want to run containers securely or insecurely until that proposal is implemented. This step would be simple I think, it is just a matter of passing variables when creating the containers on the other hosts.

**Comment by knick-burns** [comment]
It surprises me that security is thrown around as the reason why this has not been implemented, yet there is no inherent way to restrict egress traffic other than an external device, or by manipulating the ip tables.

I don't really need you to be concerned with how I run my applications.

**Comment by TAGC** [comment]
@purplesrl and @knick-burns sum it up.

> At this point it is not likely that --privileged will be added as-is to docker service, as this option nullifies all security that containers provide

I don't care about the security that containers provide. So what if a Docker container has unrestricted privileges when the host it's running on is a £30 throw-away Raspberry Pi? Why not just permit people to use `--privileged` and if they shoot themselves in the foot with it, that's on them?

What I actually do want is the service discovery, self-healing, self-containment and modularity you get from using Docker swarm.

**Comment by edward-of-clt** [comment]
I agree this would be a great help. Mouting NFS is impossible without hacking the stuff apart.

**Comment by cpuguy83** [comment]
@edward-of-cit you can always use a volume to mount nfs.

**Comment by edward-of-clt** [comment]
@cpuguy83 When you can point me to documentation that works for docker service and docker stack, I'll believe you. I have spent the last two days looking for NFS mount documentation and nothing I have found works.

**Comment by cpuguy83** [comment]
@edward-of-clt In addition to the built-in support for low-level NFS options, there are a number of volume drivers out there as well.

For the built-in support:

```yaml
services:
    foo:
        mounts:
            - type: volume
              source: mynfs
              target: /data

volumes:
    mynfs:
        driver_opts:
            type: "nfs"
            o: addr=<nfs server addr>
            device: :/path/to/share
```

**Comment by edward-of-clt** [comment]
See, I tried that. It's not working. Nothing ends up being mounted at the mount point.

And this is what I always get:

mounts Additional property mounts is not allowed

**Comment by cpuguy83** [comment]
Probably need to set the "version" at the top of the stack file to 3.1 or 3.2... Can't remember what version it came out in. 3.4 is what's included with 17.09.

**Comment by edward-of-clt** [comment]
That's not working either.

**Comment by cpuguy83** [comment]
Happy to help on the community slack.

**Comment by edward-of-clt** [comment]
I'm on there, but can't see anything but the messages that have been sent to me.

**Comment by purplesrl** [comment]
@edward-of-clt 

From what I have researched you can use this:

https://github.com/mavenugo/swarm-exec

To execute a command across the swarm with privileged flag. There is also a tutorial for what you need here:

https://www.vip-consult.solutions/post/persistent-storage-docker-swarm-nfs

**Comment by cpuguy83** [comment]
We talked offline. The linked items are really not what you want for mounting NFS... but in the interest of keeping this issue from diverging if anyone would like to discuss this, feel free to ping me on slack.

**Comment by efocht** [comment]
Our use case is that we need to assign fixed IPs to services over macvlan. We could re-assign the IPs from inside the container if we had --privileged or proper --cap-add. Nicer would be --ip option passed to the service definition similar to the way it is done with docker run.

The problem with macvlan is that it is unusable, when starting two services on two different worker nodes, we get the same IP for both.

**Comment by blop** [comment]
@efocht actually, you need to split the subnet of the swarmed macvlan for each worker
its seems that the ip allocation for macvlan networks are not distributed at the moment.

I do something like (splitting /16 into sub /24 per worker) : 

```
docker network create --config-only --subnet 10.140.0.0/16 --gateway 10.140.0.1 -o parent=ens160 --ip-range 10.140.1.0/24 local-network-name
docker network create --config-only --subnet 10.140.0.0/16 --gateway 10.140.0.1 -o parent=ens160 --ip-range 10.140.2.0/24 local-network-name
docker network create --config-only --subnet 10.140.0.0/16 --gateway 10.140.0.1 -o parent=ens160 --ip-range 10.140.3.0/24 local-network-name

docker network create -d macvlan --scope swarm --attachable --config-from local-taktik swarm-network-name
```

**Comment by efocht** [comment]
The setup above won't allow the failover a service with a particular IP. For example a user-space NFS server. I need service1 to have ip1, service2: ip2. And keep the IPs when the services are failed over.

**Comment by matthanley** [comment]
@efocht could you decouple the IP addressing from the container networking in your environment? In our environment we run keepalived on the hosts to provide floating IP addresses and then tcp-proxy back into the container networks. This won’t work if you need to dynamically assign IP addresses but you can proxy to different containers based on port instead of address.

**Comment by efocht** [comment]
@matthanley thank you for the hint, we will consider this solution. I'm a bit reluctant to add another high availability layer, though. Swarm's capabilities would actually be sufficient.

I don't want to hijack the thread to solve my issue, just wanted to report that we also have a use case for the addition of --cap-add option to services and hope that the July 20 proposal from @thaJeztah will be implemented and pushed soon.

**Comment by cpuguy83** [comment]
I don't think we'll be adding `--privileged` or `--cap-add`. See #32801 for the current way we want to solve this both for swarm and k8s. There was a talk on this at DockerCon EU... struggling to find the video at the moment.

**Comment by cpuguy83** [comment]
Actually I guess the talk was at Moby Summit. These videos have not been posted yet.

**Comment by mwaeckerlin** [comment]
I need the option `--priviledged` for running jenkins in a swarm. Jenkins in my configuration needs to run docker within docker, because all builds are started in a fresh docker container.

See https://github.com/mwaeckerlin/jenkins

Advantages:
 - builds are highly reproducible
 - build packages for all kind of Linux distributions and version, by simply running a container of that version
 - no need to install build tool and library dependencies in the main jenkins container / service, but install them in the build container (or store images that already contain the dependencies)

So for me, running docker in docker is an absolut must requirement to be able to migrate jenkins from current single host local docker to my new docker swarm environment.

AFAIK, running docker in docker requires `--priviledged`, so for me, `--priviledged` is also an absolute must requirement to docker swarm, unless you can show me another solution for running docker containers in docker services.

**Comment by jonasddev** [comment]
@mwaeckerlin I'm don't have the details of what you're trying to do, but I'm thinking that you may not need docker in docker, but start sibling dockers from docker by mounting the docker socket as a volume.

Have a read at this: https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/

**Comment by mwaeckerlin** [comment]
@jonasdddev, well I don't really bother the «bad», «ugly» and «worse», because I have a working setup. ;)

But I'll give that solution mentioned there a try and just bind-mount `/var/run/docker.sock` into the jenkins image.

**Comment by jonasddev** [comment]
@mwaeckerlin Dito, don't care about it either, but was in your situation even before Swarm and that solution solved all my perceived problems. Just let me know for my info if it was successful.

Thanks.

**Comment by tk0485** [comment]
I have same use case as @man4j. I want to use gluster within swarm and I need priviillaged= true, is there any work around to make it happen? I get "setting extended attributes" error

**Comment by man4j** [comment]
Hi all. I have a workaround. I create a swarm service from my image with next entrypoint:
#!/bin/bash
exec docker run --init --rm --privileged=true --cap-add=ALL ...etc

It's all. When service tasks are created they run my privileged container and when tasks are destroyed SIGTERM kills child privileged container thanks to --init flag.

And of course we need to map docker.sock and install docker inside image for parent service.

Dockerfile:

FROM docker
RUN apk --no-cache add bash
COPY entrypoint.sh /
RUN chmod +x /entrypoint.sh
ENTRYPOINT ["/entrypoint.sh"]

entrypoint.sh:
#!/bin/bash
exec docker run --init --rm --privileged=true --cap-add=ALL ...etc

service-run.sh:
#!/bin/bash
docker service create --name myservice --detach=false \
--mount "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock" myimage

Waiting for your thoughts!

**Comment by matthanley** [comment]
@man4j this is the same workaround @port22 mentioned in https://github.com/moby/moby/issues/24862#issuecomment-326124626

**Comment by mwaeckerlin** [comment]
Just mounting `/var/run/docker.sock` as volume does not work, @jonasddev. Just because the group id of group docker is different inside the container than outside, so if the user inside the container is not root, but only member of group docker, as it should be, then it has no access to the socket!

Any good solution for this?

**Comment by ulm0** [comment]
I'm facing this issue as well when trying to deploy GitLab in Docker Swarm

The package *needs* to set values but it can't and the service just keep restarting

```sh
/opt/gitlab/embedded/bin/runsvdir-start: line 24: ulimit: pending signals: cannot modify limit: Operation not permitted
/opt/gitlab/embedded/bin/runsvdir-start: line 37: /proc/sys/fs/file-max: Read-only file system
```

**Comment by mwaeckerlin** [comment]
@jonasddev, I updated my jenkins image to fix the permissions within the container, I added to my entry-point-script:
```
# add user to group that has access to /var/run/docker.sock
addgroup --gid $(stat -c '%g' /var/run/docker.sock) extdock || true
usermod -a -G $(stat -c '%g' /var/run/docker.sock) jenkins || true
```

Now the next problem:
 - docker in docker, access through shared `/var/run/docker.sock`
 - mounting volumes does not work

This is the problem in a `mwaeckerlin/jenkins` container:
```
ubuntu$ docker run -d --restart unless-stopped -v /var/run/docker.sock:/var/run/docker.sock --name jenkins -p 8080:8080/tcp -p 50000:50000/tcp --volumes-from jenkins-volumes mwaeckerlin/jenkins
ubuntu$ docker exec -it -u jenkins jenkins bash
jenkins$ docker create -v /var/lib/jenkins/workspace/mrw-c++.rpm/distro/fedora-27:/workdir -v /var/lib/jenkins/.gnupg:/var/lib/jenkins/.gnupg -e LANG=en_US.UTF-8 -e HOME=/var/lib/jenkins -e TERM=xterm -e DEBIAN_FRONTEND=noninteractive -e DEBCONF_NONINTERACTIVE_SEEN=true -e BUILD_NUMBER=64 -w /workdir fedora:27 sleep infinity
jenkins$ docker start cd93d11c2634b4e4094cc2541996e30a4c08c60923b8896e0bd7e439c7d9c673
jenkins$ ls /var/lib/jenkins/workspace/mrw-c++.rpm/distro/fedora-27
AUTHORS                      COPYING                   mrw-c++.spec.in
autogen.sh                   debian                    NEWS
ax_check_qt.m4               demangle.h                README
ax_cxx_compile_stdcxx_11.m4  dependency-graph.sh       resolve-debbuilddeps.sh
ax_init_standard_project.m4  doc                       resolve-rpmbuilddeps.sh
bootstrap.sh                 examples                  rpmsign.exp
build-in-docker.conf         INSTALL                   sql-to-dot.sed
build-in-docker.sh           mac-create-app-bundle.sh  src
build-resource-file.sh       makefile.am               suppressions.valgrind
ChangeLog                    makefile_test.inc.am      template.sh
checkinstall.sh              mrw-c++.desktop.in        test
configure.ac                 mrw-c++-minimal.spec.in   valcheck.sh
jenkins$ docker exec -u 107 -it 033088f9008601d5f9f9034744b579910ee1f88bdf3a61d2f0354b8454ba94ed bash
docker$ ls /workdir
docker$ mount | grep /var/lib/jenkins/workspace/mrw-c++.rpm/distro/fedora-27
/dev/mapper/big-root_crypt on /workdir type btrfs (rw,relatime,space_cache,subvolid=257,subvol=/@/var/lib/jenkins/workspace/mrw-c++.rpm/distro/fedora-27)
docker$
```

So in a jenkins-container (here named `jenkins`, a container (here named `docker`) is started. A directory from the jenkins comtainer is mounted int the container in the container, but there, it is empty. The mount is visible in the `mount` command.

Any idea?

**Comment by mwaeckerlin** [comment]
BTW: #21109

**Comment by mwaeckerlin** [comment]
The problem is describe here in a coment: http://container-solutions.com/running-docker-in-jenkins-in-docker/
> When using docker run inside the jenkins container with volumes, you are actually sharing a folder of the host, not a folder within the jenkins container. To make that folder “visible” to jenkins (otherwise it is out of your control), that location should have a parent location that matches the volume that was used to run the jenkins image itself.

Of course, I want to mount from docker container to the container in the container and not from the host.

Any simple solution for this?

**Comment by mwaeckerlin** [comment]
# Security Problem

If using  `-v /var/run/docker.sock:/var/run/docker.sock` for docker in docker, you get full access to the host and you are no more jailed within the container!

Here a demonstration of the security issue mounting the socket:

```
marc@jupiter:~/docker/dockindock$ echo 'this is the host' > /tmp/test
marc@jupiter:~/docker/dockindock$ docker run -d --name dockindock -v /var/run/docker.sock:/var/run/docker.sock -v $(which docker):/usr/bin/docker mwaeckerlin/dockindock sleep infinity
d44fbd58e44a180e388621d39aff64652bc4118973c2cbc86a1738ffb481ebaf
marc@jupiter:~/docker/dockindock$ docker exec -it dockindock bash
root@docker[d44fbd58e44a]:/# cat /tmp/test
cat: /tmp/test: No such file or directory
root@docker[d44fbd58e44a]:/# echo 'this is the outer container' > /tmp/test
root@docker[d44fbd58e44a]:/# docker ps
CONTAINER ID        IMAGE                    COMMAND             CREATED             STATUS              PORTS               NAMES
d44fbd58e44a        mwaeckerlin/dockindock   "sleep infinity"    16 minutes ago      Up 16 minutes                           dockindock
root@docker[d44fbd58e44a]:/# docker run -it --rm -v /tmp:/tmp mwaeckerlin/ubuntu-base bash
root@docker[f197131c54c7]:/# cat /tmp/test
this is the host
```

So, the solution of using `-v /var/run/docker.sock:/var/run/docker.sock` is a security issue and an **absolute no-go!** @jonasddev 

So either we get `--privileged` for swarm, or there is need for any other, better solution, that solves this issue!

# Wanted Behaviour

Here a demonstration of how it should work, see the restricted access:
```
marc@jupiter:~/docker/dockindock$ echo 'this is the host' > /tmp/test
marc@jupiter:~/docker/dockindock$ docker run -d --name dockindock --privileged mwaeckerlin/dockindock
655c67b2a6d9f06da8bf630889710ee596f006331a036dc009c86a9e04ea0201
marc@jupiter:~/docker/dockindock$ docker exec -it dockindock bash
root@docker[655c67b2a6d9]:/# cat /tmp/test
cat: /tmp/test: No such file or directory
root@docker[655c67b2a6d9]:/# echo 'this is the outer container' > /tmp/test
root@docker[655c67b2a6d9]:/# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
root@docker[655c67b2a6d9]:/# docker run -it --rm -v /tmp:/tmp mwaeckerlin/ubuntu-base bash
Unable to find image 'mwaeckerlin/ubuntu-base:latest' locally
latest: Pulling from mwaeckerlin/ubuntu-base
1be7f2b886e8: Pull complete 
6fbc4a21b806: Pull complete 
c71a6f8e1378: Pull complete 
4be3072e5a37: Pull complete 
06c6d2f59700: Pull complete 
04fca7013ee9: Pull complete 
7a66494bf7fe: Pull complete 
be1530d02718: Pull complete 
57cb4fb92cd1: Pull complete 
4170a785b84a: Pull complete 
36570a7926c8: Pull complete 
34218f1ce9d6: Pull complete 
Digest: sha256:e9207a59d15739dec5d1b55412f15c0661383fad23f8d2914b7b688d193c0871
Status: Downloaded newer image for mwaeckerlin/ubuntu-base:latest
root@docker[bcfc1e9bc756]:/# cat /tmp/test
this is the outer container
```

**Comment by bitsofinfo** [comment]
Please add, needed for fuse!

**Comment by thaJeztah** [comment]
> So, the solution of using -v /var/run/docker.sock:/var/run/docker.sock is a security issue and an absolute no-go!

Correct; if you’re blind-mounting the socket, you’re not running docker-in-docker; you’re controlling the hosts daemon from inside the container; in many cases this may actually be preferable, see http://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/

> The option --privileged is a must to have docker containers in docker containers, encapsulated without access to the whole swarm!

First of all, only _manager_ nodes have access to the whole swarm; worker nodes can only control the worker node itself

But be aware that `—privileged` is equivalent to having full root access on the host; there is _no_ protection whatsoever, processes inside the container can escape the container, and have full access to the host (for example, have a look at `/dev` inside a privileged container, and you see it has access to all devices from the host)

**Comment by mwaeckerlin** [comment]
@thaJeztah writes:
> But be aware that —privileged is equivalent to having full root access on the host; there is no protection whatsoever, processes inside the container can escape the container, and have full access to the host (for example, have a look at /dev inside a privileged container, and you see it has access to all devices from the host)
Then let's adapt the requirement. What is a good solution to have:
 - Real independent Docker in Docker
 - Full security, no access to the host, no escape
 - Can run in a swarm

For me, not the `--privileged` flag is important, but the real and secure docker in docker.

Possible use cases:
 1. Jenkins with untrusted users that can setup their own build jobs
 1. Sell docker playground to any unkown audience, let them start any docker services from within docker

**Comment by AkihiroSuda** [comment]
@mwaeckerlin 

We need to get nested runc to work without root privileges first.
You may want to follow this issue: https://github.com/opencontainers/runc/issues/1658

**Comment by redhog** [comment]
I'd like to add that Kubernetes does support adding capabilities. For an example, check out https://caveofcode.com/2017/06/how-to-setup-a-vpn-connection-from-inside-a-pod-in-kubernetes/ (this is the exact same use case I have, and that I'd like to run on docker swarm).

**Comment by thiagolsfortunato** [comment]
Any news ?

**Comment by bitsofinfo** [comment]
the news is all k8!

**Comment by denis-isaev** [comment]
+1

**Comment by killcity** [comment]
Gave up on this and switched to Kubernetes.

**Comment by kurtrwall** [comment]
Just first off, thank you @thaJeztah and co-devs for being attentive on this. I'm sure this is a point of contention among yourselves.

Really just came here to say that I've been successful with bind mounting the docker socket and using docker binaries in the container with Jenkins in a swarm (not bind mounting the binaries). In that specific case, I've constrained Jenkins to only run on manager nodes. Anyone should make sure the permissions are set correctly, especially when you're working with GlusterFS, as all of my Jenkins nodes run on top of GlusterFS volumes with no issues so far (been about 6 months).

The `--privileged` flag was something I used when prototyping but when it wasn't available I found that just setting the correct permissions did the trick for my purposes; something everyone should be aware of and practicing when bind mounting anyway. My case is different than many others, so your mileage will vary.

**Comment by TAGC** [comment]
Damn, forgot to come back and congratulate this issue on it's second birthday.

**Comment by denis-isaev** [comment]
[zabbix-agent](https://hub.docker.com/r/zabbix/zabbix-agent/) need privileged mode to monitor host resources.

**Comment by mister2d** [comment]
Need privileged mode to run systemd enabled containers! Like dogtag-ca for instance.

**Comment by valleedelisle** [comment]
Trying to deploy [dell openmanage exporter](https://hub.docker.com/r/galexrt/dellhw_exporter/) and it doesn't work. I understand that --priv / --cap-add can be a security issue, but if I want to shoot myself in the foot, I should be able to do so.

**Comment by manvalls** [comment]
Hey guys, I've been maintaining my own poorly-patched fork of docker for a while with support for `--cap-add` and `--privileged` in `docker stack deploy`, maybe we should create a proper fork?

**Comment by thaJeztah** [comment]
@manvalls I think `--cap-add` (or `--capabilities`, i.e., not merging defaults, but require the full set to be specified) is something that would be accepted; have you considered contributing, and opening a pull request to discuss that option?

**Comment by olljanat** [comment]
@manvalls swarmkit team have made proposal of device support to docker/swarmkit#2682

Please comment to there your thought about if that fits to your use cases.

**Comment by thaJeztah** [comment]
@manvalls the PR you linked to was closed because that implementation did `--cap-add` / `--cap-drop`. The proposal was to have a `--capabilities` flag (which would override the defaults) 
https://github.com/moby/moby/pull/26849#issuecomment-249176129 :

>  I think it is better long term if we switch to a `--capabilities` interface which says which you actually need, even if you might get more by default for compatibility.

If someone wants to work on that, that's a change that will likely get accepted.

**Comment by pwFoo** [comment]
Also need device, privileged, cap_add / cap_drop. No progress here?

**Comment by AkihiroSuda** [comment]
@pwFoo A workaround is to bind-mount `/var/run/docker.sock` and call `docker run` within service containers.

Maintainers: Although I agree we should support "entitlements" in the long-term plan, as we already have been supporting bind-mounting arbitrary path including `docker.sock`, I don't see any security degradation in implementing `docker service create --privileged`.

**Comment by olljanat** [comment]
@pwFoo there is that #26849 PR which you can take as template and just implement asked changes to get it merged (look comments above). That beauty of open source that you can do that yourself 😉

It is still possible to get that feature out on 19.03 if someone just take to step and start implement it.

**Comment by pwFoo** [comment]
I'm new with Go. So I don't think I should do it...
But maybe someone with more experience could do it.

**Comment by mhemrg** [comment]
What's the state of this issue?

**Comment by AkihiroSuda** [comment]
needs somebody to take over https://github.com/moby/moby/pull/26849

**Comment by olljanat** [comment]
FYI. I started implementing capabilities feature. You can see status on https://github.com/moby/moby/issues/25885#issuecomment-447657852

**Comment by aderbique** [comment]
Pleeeeaaaaase add this feature in :(

**Comment by olljanat** [comment]
Pleeeeaaaaase don't send useless messages here. It just slowdown the process as lot of people get notified and that time is away from actual implementation.

My plan is to get #38380 released as part of 19.03 and then actual Swarm side changes on version which comes after that (18.06 I guess). Anyone who want to help plz test that and tell comments on PR so then hopefully we can stay on that schedule.

**Comment by ntimo** [comment]
This issue is open since 2016 is there a chance that this gets added to Docker in the next years.... ? This is a must have feature, for running Docker in Docker on a swarm. Please Docker team.

Sorry but this is very frustrating for me. I can't switch to a HA setup because of this :(

**Comment by olljanat** [comment]
#39173 was merged so this feature will ship as part of Docker 19.06 / 19.09 (not sure actual version yet).

Please, look my question about CLI side implementation on https://github.com/moby/moby/issues/25885#issuecomment-501017588
and here you can also comment use cases which actually would need `--privileged` switch or can all those be handled by defining all needed capabilities? (nothing prevents user to list all of them)

**Comment by olljanat** [comment]
FYI https://github.com/moby/moby/issues/25885#issuecomment-557790402

**Comment by SySfRaMe** [comment]
Hi
I to seek for the possiblity to attach a device such as the /dev/ttyACM0 to a stack deployed service.
I created a script to run based on crontab every 5 min. (on the host of every node)
If the script see´s my special device attached. Then it will raise a node flag.
This flag is then used as a constraint to what node the service is allowed:

If lets say node1 failes then i can move the usb-dongle to node2 and the service is started there.


If you are interested in the cron script then look here:
https://github.com/SySfRaMe/Docker_ZwaveHW_Flag

**Comment by kaysond** [comment]
I am trying to run `ip netns exec <namespace> iptables -L` from within a swarm container, and it seems that without `--privileged`, I'm unable to do so. I'm guessing its because `/proc`, `/sys`, `cgroup` etc are mounted as RO. Does anyone know of a workaround for this?

If not, it seems my container must have `--privileged` in swarm.

**Comment by olljanat** [comment]
Note on Docker 20.10.x you can use `--cap-add` parameter with swarm services. `--privileged` switch still is not available (and will no be added because it makes too easy to create unsecure configuration) but nothing prevents you to list all capabilities you need.

To allow iptables comnand inside of container you need to use `--cap-add NET_ADMIN`

@thaJeztah I think that we can actually close this issue?

**Comment by kaysond** [comment]
@olljanat the problem in my case is different. i already have `--cap-add NET_ADMIN`, which allows me to run `iptables`. the problem is `ip netns exec` runs `mount --make-rslave /`, which fails due to the read-only `/proc` `/sys`, etc mounts. so i can only run iptables on the default namespace

**Comment by olljanat** [comment]
@kaysond on that case plz create new issue with all asked details.

**Comment by beornf** [comment]
@kaysond I also want to run `ip netns exec` scoped commands in a swarm container. A more direct way to make the mounts readable would be to add the parameter `--security-opt apparmor=unconfined`. There's an open issue #41371 for adding the `--security-opt` option for services.

**Comment by kaysond** [comment]
@beornf a workaround that I was able to use is `nsenter --net=/var/run/netns/<namespace> <command>`. I used host networking and set up a bind mount: `/var/run/docker/netns:/var/run/netns`, then used `NET_ADMIN` and `SYS_ADMIN` `cap_add`'s on the service.

It similarly runs a process in the specified namespace, it just doesn't do all of the remounting that `ip netns exec` does. Luckily, it doesn't seem that many of the stock linux networking utilities actually need that remounting.

**Comment by dimisjim** [comment]
> @beornf a workaround that I was able to use is `nsenter --net=/var/run/netns/<namespace> <command>`. I used host networking and set up a bind mount: `/var/run/docker/netns:/var/run/netns`, then used `NET_ADMIN` and `SYS_ADMIN` `cap_add`'s on the service.
> 
> It similarly runs a process in the specified namespace, it just doesn't do all of the remounting that `ip netns exec` does. Luckily, it doesn't seem that many of the stock linux networking utilities actually need that remounting.

Could you provide a sample compose yaml demonstrating this?

**Comment by kaysond** [comment]
> Could you provide a sample compose yaml demonstrating this?

It's just

```
volumes:
  - /var/run/docker/netns:/var/run/netns
```
